### DISCUSSION TITLE
How to keep clickhouse easily in sync with mysql?

### DISCUSSION THREAD

**Discussion by rick20** [ClickHouse/ClickHouse#44887] [Q&A]
What is the most simple way to keep clickhouse always in sync with mysql where both are put in a single server? no cluster, no mirroring/replication.
Many tutorials use complicated technique to achieve that using debezium, kafka and other.
While it may necessary for large multicluster or distributed data store, our case is much more simple.

Currently we frequently do bulk insert, update, or delete data in mysql.
We need to do these operations in clickhouse as well, but the docs says that the update/delete operations are not intended for frequent use.
So the mysql tables are still needed, but how do we keep both in sync?

Thinking about truncating the clickhouse tables and re-insert all data from mysql each time a mutation is happened in mysql side.
But it doesn't sound a good idea when working with large data.

So I'm looking for advice here, I'd really appreciate that :)

**Comment by tsolodov** [comment]
You have few options:
- try this experimental feature: https://clickhouse.com/docs/en/engines/database-engines/materialized-mysql/
- you can use MySQL table Engine: https://clickhouse.com/docs/en/engines/table-engines/integrations/mysql/

**Comment by geopet85** [comment]
The only production ready method is to use debezium/kafka and either pull them into Clickhouse using a Kafka engine, or push them via a kafka connector (all options available [here](https://clickhouse.com/docs/en/integrations/kafka)). In both cases, you would need additional materialized view to parse debezium messages and eventually tables  with `ReplacingMergeTree` engine to handle the inserts, updates and deletes.

**Comment by aadant** [comment]
There is this project too now v2.0 and GA https://github.com/Altinity/clickhouse-sink-connector
We use it to replicate TB scale MySQL without issues.

It handles most DDL (sink-connector-lightweight)

**Reply by raghukn** [reply]
Does it use KafakEngine Table and ReplacingMergeTree on CH side to keep data rows with out duplication?? Is this a better version than coding KafkaEngine for each Topic and consuming in a ReplacingMergeTree MV?

**Comment by bakwc** [comment]
MaterializedMySQL has several major limitations:

1. It does not removes data, the storage keeps growing.
2. No support for bools, jsons and more. No easy way to add what's missing.
3. If you have multiple DBs - network traffic growing exponentially.
4. Sometimes it crashes, hard to debug and fix C++ code.
5. Long downtimes when restarting replication.

As an alternative I would suggest to look into https://github.com/bakwc/mysql_ch_replicator.
 - it's python (with cpp extension for performance), all above issues resolved, easy to add if anything missing
 - no need of kafka, 3d party connectors, etc, it's a single standalone app
